### ISSUE TITLE
List of objects in a dir

### ISSUE THREAD

**Issue by iwanbk** [zero-os/0-stor#502]
S3 and Minio has feature to list objects in a bucket (namespace in 0-stor term).
arguments:
- prefix : prefix of the object name
- max objects : max number of objects to show

Use case:
- browse the objects on a bucket  (using browser or another ftp-like client)
- before deleting the object

**Comment by iwanbk** [comment]
> Other than to browse the objects on a bucket (using browser or another ftp-like client), i don't see other use cases that need this feature.

After more tests, minio also use this feature upon object deletion. It failed when we don't have this feature.

**Comment by GlenDC** [comment]
We'll have to think about how to implement this feature. We can already list objects on a datastor server, but obviously that would require in that case for you to know all the servers you want to check. So as long as that is possible, it shouldn't be an issue, as we can combine the search results of multiple datastor servers. It will however give a lot of duplicates. As the same object can be spread over multiple (datastor) shards.

However, to just know a namespace, without knowing all datastor servers, it isn't possible at the moment to do so. If somehow we store a list of object keys, for each namespace, we could do it, but that seems pretty bad, and error prone to manage.

So yeah, should this feature be required, we'll have to design and plan it well.

**Comment by zaibon** [comment]
One we have the 0-metadata project going on, we can replace etcd with it.
And since it will use a directory structure we can map a directory to a namspace and do a listing of all the files in a namespace together with the metadata

**Comment by GlenDC** [comment]
Let's talk on Monday also a bit about 0-metadata @zaibon, as I right now don't know anything about it.

**Comment by GlenDC** [comment]
Now that 0-metadata is delayed, @zaibon, what should we do with this?

As noted in https://github.com/zero-os/0-stor/issues/502#issuecomment-362738469, we can already list objects for a namespace on a single zstordb server, however, there is currently no way to do the same for all "known" objects.

One way could be to maintaining a list in the metadata of all objects (files/) in a namespace, as you already proposed. But I guess that doesn't have to wait on 0-metadata, we could just extend the API and internal logic to allow for this?

**Comment by zaibon** [comment]
Yes indeed that's also how I was seeing things. We could extend the metastor api to allow listing of all files from a namespace.

**Comment by GlenDC** [comment]
If so, why do we close this issue already, that work still has to be done right?

**Comment by zaibon** [comment]
probably a miss click I didn't wanted to close this.

**Comment by iwanbk** [comment]
For minio case, it  mimic this S3 API https://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html

As a quick explanation, `prefix` header there can be used to mimic a directory.
Example:
- to create file in dir `/a/b`, we use this key `/a/b/filename`
- to list all files in dir `/a/b`, we use `prefix`=`/a/b` in list objects API

**Comment by zaibon** [comment]
As far as I see it, this doesn't map with our namespace concept, cause I don't think we support multiple level of namespaces.
From what I see we won't even be able to use the badger metastor as it is now either. Because the key used in the Metadata structure is just the plain key of the file itself, we don't have any notion of namespace. 

Maybe what we could do it to add the namespace to the metadata, so the `SetMetadata` method of the metastor client could organize the metadata per namespace so we could list the full content of it

**Comment by iwanbk** [comment]
> Add a List(namespace []byte), metadatas [][]byte, err error) to the metastor.db.DB interface so we can list all object from a namespace.

I propose to add:
- prefix filter : inspired by S3 API, useful if we manage our keys in dir like structure
- paging  mechanism: in case we have a lot objects

**Comment by zaibon** [comment]
We should indeed probably provide an interface that let us paginate.
We can use the same idea as we do in the datastor, something like:
`List(ctx context.Context) (<-chan []byte, error)`
So the consumer can paginate as he wants just by consuming the channel.

Regarding prefix, maybe, I'm not sure, but what I proposed is already this idea, just with another name. But allowing prefix means we would could store anything on the metadatastor. Which is maybe something we would like to support in the futur for custom metadata.

**Comment by GlenDC** [comment]
> `List(ctx context.Context) (<-chan []byte, error)`

I would define the signature of the new `metastor/db.DB` interface method `List` as follows:
```go
List(cb func(key []byte) error, namespace []byte) error
```

Should one want to transfer it over a goroutine, it's easy to do by simply spawning one. But at least if you are only wanting to use it synchronously you don't waste time on transporting it over a channel for nothing. On top of that it's also way less error prone to use, which is a minor point, but using the channel approach one has to not forget to mark the context as done, in case the user wishes to loop early, and no tooling I know of will warn you if you forgot  doing that.

> Regarding prefix, maybe, I'm not sure, but what I proposed is already this idea, just with another name. But allowing prefix means we would could store anything on the metadatastor.

Fairly sure you'll have to start prefixing your keys (using the given namespace), I don't see how else you'll efficiently list keys based on a namespace.

> So the consumer can paginate as he wants just by consuming the channel.

This would be true for badger, but if the metastor is something like ETCD, you would probably also want to start paging on the metastor database level, otherwise the paging on the higher level wouldn't really help a lot, no?

> prefix filter : inspired by S3 API, useful if we manage our keys in dir like structure

Ideally we leave it up to the database implementation how to structure keys and values, shouldn't really be something a higher level need to worry about.

Regarding the prefix, it seems we just make it only work within a namespace. You could instead make the metastor client and database work with a prefix instead, which would mean you only really have to add one function `List(cb func(key []byte) error, prefix []byte) error` (where the `prefix` parameter could be optional), as in that design the other functions don't need any change, as the prefix would simply be already part of the key prior to calling those DB functions (e.g. `Set`/`Get`/`Delete`...).

But really you'll have to agree on what kind of feature you need, either going for the namespaced keys, or simply allowing to list on a prefix, and let the prefixing be the responsibility of the `metastor.Client`.

**Comment by iwanbk** [comment]
> So the consumer can paginate as he wants just by consuming the channel.

While it is an idiomatic solution, also works well with grpc.
It is worth to note that it doesn't really help when it is being used in stateless REST server.

**Comment by GlenDC** [comment]
>  it is an idiomatic solution

It's an idiomatic solution, but not the most performant solution, neither the most user-friendly one. See https://github.com/zero-os/0-stor/issues/502#issuecomment-366889927.

For some background you can check blog posts like https://ewencp.org/blog/golang-iterators/.

**Comment by zaibon** [comment]
I'm also fine with `List(cb func(key []byte) error, namespace []byte) error`, since we also use this construct in the datastor.DB.

> Fairly sure you'll have to start prefixing your keys (using the given namespace), I don't see how else you'll efficiently list keys based on a namespace.

Yes that's what I'm saying. To allow listing of key in a namespace, you must add a prefix to your key, or any other way of segragation provided but the database.

> But really you'll have to agree on what kind of feature you need, either going for the namespaced keys, or simply allowing to list on a prefix, and let the prefixing be the responsibility of the metastor.Client.

low level need to be generic, so we'll go for prefix and metastor.Client will use the namespace as prefix key.

**Comment by iwanbk** [comment]
I can't see how etcd & badger could help us to  provide dir-like listing to the `List` API.
Lets say we have these keys:
- a/b
- a/b/d
- a/c

When we list keys with `a/` as prefix, we'll get `a/b/d` as well.

Of course we could filter it  by ourselves, but it is not efficient.

> This would be true for badger, but if the metastor is something like ETCD, you would probably also want to start paging on the metastor database level, otherwise the paging on the higher level wouldn't really help a lot, no?

For etcd, it seems we can provide paging mechanism with combination of [WithFirstKey](https://godoc.org/github.com/coreos/etcd/clientv3#WithFirstKey) and [WithLimit](https://godoc.org/github.com/coreos/etcd/clientv3#WithLimit)

As for badger, [Iterator.Seek](https://godoc.org/github.com/dgraph-io/badger#Iterator.Seek) could help us.

I propose to add `prefix` and  `maxKeys` arguments to the `List` API, both are used for paging.

```go
List(cb func(key []byte) error, namespace, prefix []byte, maxKeys int ) error
```

**Comment by zaibon** [comment]
Let's see if we will need it. For first version let's implement without paging.

**Comment by iwanbk** [comment]
hi guys, i think i make stupid mistake in my code when implementing `List` API and stuck with it.

The code snippet
```go
var (
        strKeys  []string
        byteKeys [][]byte
        mux      sync.Mutex
    )
    // list keys
    err := db.List(namespace, func(key []byte) error {
        mux.Lock()
        // print the key we got and content of byteKeys
        fmt.Printf("got= %v, byteKeys=%v\n", string(key), fmtByteKeys(byteKeys))

       // append key to byte keys
        byteKeys = append(byteKeys, key)

        // append key to stringKeys
        strKeys = append(strKeys, string(key))

        // print byteKeys content
        fmt.Printf("\t byteKeys = %v\n", fmtByteKeys(byteKeys))

       // print string keys content
        fmt.Printf("\t strKeys=%v\n", strKeys)
        mux.Unlock()
        return nil
    })
```

What the above callback do is simply appending the key to slice.
The problem is : it doesn't work correctly when using `[][]byte` to store the keys. But perfectly fine with `[]string`.
If there is something wrong with `List` code, i think it should also failed with `[]string`

The output
```bash
# go test -run TestBadgerDB_List
got= mykey_0, byteKeys=
         byteKeys = mykey_0
         strKeys=[mykey_0]
got= mykey_1, byteKeys=mykey_2
         byteKeys = mykey_2 mykey_1
         strKeys=[mykey_0 mykey_1]
got= mykey_2, byteKeys=mykey_2 mykey_3
         byteKeys = mykey_2 mykey_3 mykey_2
         strKeys=[mykey_0 mykey_1 mykey_2]
got= mykey_3, byteKeys=mykey_4 mykey_3 mykey_4
         byteKeys = mykey_4 mykey_3 mykey_4 mykey_3
         strKeys=[mykey_0 mykey_1 mykey_2 mykey_3]
got= mykey_4, byteKeys=mykey_4 mykey_3 mykey_4 mykey_3
         byteKeys = mykey_4 mykey_3 mykey_4 mykey_3 mykey_4
         strKeys=[mykey_0 mykey_1 mykey_2 mykey_3 mykey_4]
```

rundown
first key
```
got= mykey_0, byteKeys=
         byteKeys = mykey_0
         strKeys=[mykey_0]
```
new key = mykey_0, all are fine:
- `byteKeys` initially still empty
-  after append, `byteKeys` and `strKeys` has correct content

Second key
```
got= mykey_1, byteKeys=mykey_2
         byteKeys = mykey_2 mykey_1
         strKeys=[mykey_0 mykey_1]
```
new key  = mykey_1
initial values of byteKeys already corrupted
after append, strKeys still fine

In case someone want to check
https://github.com/zero-os/0-stor/tree/broken-list-keys/client/metastor/db/badger

you could run `go test -run TestBadgerDB_List` from that dir

**Comment by zaibon** [comment]
I don't think you can just append the []byte into your [][]byte, you need to do a copy cause otherwise the underlying array is share and not copied

**Comment by iwanbk** [comment]
> You need to do a copy of the iterator key before passing it to the callback, this is documented by the badger iterator.

My bad, thx @zaibon !

**Comment by iwanbk** [comment]
Closed in #519
